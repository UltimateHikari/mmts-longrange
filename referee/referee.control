comment = 'Multimaster referee'
default_version = '1.0'
module_pathname = '$libdir/referee'
schema = referee
relocatable = false
