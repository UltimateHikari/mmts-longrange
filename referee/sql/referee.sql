CREATE EXTENSION referee;

SELECT * FROM referee.decision;

SELECT referee.request_grant(1, 7);
-- node can get its grant reissued
SELECT referee.request_grant(1, 9);
-- but another can't get it while the previous is not cleared
SELECT referee.request_grant(2, 4);
SELECT referee.request_grant(2, 10);
SELECT * FROM referee.decision;

DELETE FROM referee.decision WHERE gen_num < 8 OR (node_id = 1 AND gen_num <= 9);
-- surely 2 node can acquire the grant after removal of the old one
SELECT referee.request_grant(2, 11);
SELECT * FROM referee.decision;
