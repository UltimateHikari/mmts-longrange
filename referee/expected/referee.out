CREATE EXTENSION referee;
SELECT * FROM referee.decision;
 key | node_id | gen_num 
-----+---------+---------
(0 rows)

SELECT referee.request_grant(1, 7);
 request_grant 
---------------
 
(1 row)

-- node can get its grant reissued
SELECT referee.request_grant(1, 9);
 request_grant 
---------------
 
(1 row)

-- but another can't get it while the previous is not cleared
SELECT referee.request_grant(2, 4);
ERROR:  grant was already issued to node 1 in generation 9
CONTEXT:  PL/pgSQL function request_grant(integer,bigint) line 19 at RAISE
SELECT referee.request_grant(2, 10);
ERROR:  grant was already issued to node 1 in generation 9
CONTEXT:  PL/pgSQL function request_grant(integer,bigint) line 19 at RAISE
SELECT * FROM referee.decision;
  key   | node_id | gen_num 
--------+---------+---------
 winner |       1 |       9
(1 row)

DELETE FROM referee.decision WHERE gen_num < 8 OR (node_id = 1 AND gen_num <= 9);
-- surely 2 node can acquire the grant after removal of the old one
SELECT referee.request_grant(2, 11);
 request_grant 
---------------
 
(1 row)

SELECT * FROM referee.decision;
  key   | node_id | gen_num 
--------+---------+---------
 winner |       2 |      11
(1 row)

